src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs

```diff
@@ -203,36 +204,44 @@ protected void LookupMembersInternal(LookupResult result, NamespaceOrTypeSymbol
             compatibleExtensions.Free();
             return;
 
-            void getCompatibleExtensions(Binder binder, TypeSymbol type, ArrayBuilder<NamedTypeSymbol> compatibleExtensions)
+            void getCompatibleExtensions(Binder binder, TypeSymbol type, ArrayBuilder<NamedTypeSymbol> compatibleExtensions, ConsList<TypeSymbol>? basesBeingResolved)
             {
                 var extensions = ArrayBuilder<NamedTypeSymbol>.GetInstance();
                 binder.GetImplicitExtensionTypes(extensions, originalBinder: this);
 
                 foreach (var extension in extensions)
                 {
-                    // PROTOTYPE deal with generic extensions
-                    if (isCompatible(extension, type))
+                    if (basesBeingResolved?.Contains(extension) == true)
+                    {
+                        continue;
+                    }
+
+                    if (isCompatible(extension, type, out NamedTypeSymbol? substitutedExtension))
                     {
                         // PROTOTYPE should we deal with duplicate or near-duplicate extensions here instead of in merging/hiding logic?
-                        compatibleExtensions.Add(extension);
+                        compatibleExtensions.Add(substitutedExtension);
                     }
                 }
 
                 extensions.Free();
             }
 
-            bool isCompatible(TypeSymbol extension, TypeSymbol type)
+            bool isCompatible(NamedTypeSymbol extension, TypeSymbol type, [NotNullWhen(true)] out NamedTypeSymbol? substitutedExtension)
             {
                 Debug.Assert(!type.IsExtension);
                 var extended = extension.ExtendedTypeNoUseSiteDiagnostics;
                 if (extended is null)
+                {
+                    substitutedExtension = null;
                     return false;
+                }
 
                 var baseType = type;
                 do
                 {
-                    if (matches(extended, baseType))
+                    if (TypeUnification.CanImplicitlyExtend(extended, baseType, out var map))
                     {
+                        substitutedExtension = map is not null ? (NamedTypeSymbol)map.SubstituteType(extension).Type : extension;
```

Comment:
>map.SubstituteType(extension) [](http://example.com/codeflow?start=82&length=29)

Is it possible that some type parameters are left unsubstituted after that? For example when extension in metadata violates the ```ERR_UnderspecifiedImplicitExtension``` condition.

Reply:
Yes, it's possible in source. This is illustrated by `UnderlyingType_TypeParametersMustBeUsed_MissingOne`.
But I hadn't considered the bad metadata scenario. In such case, we'd end up with a partially substituted extension type but no other error...

Reply:
> But I hadn't considered the bad metadata scenario. In such case, we'd end up with a partially substituted extension type

I assume for source scenario we also were ending up with partially substituted extension type

Reply:
> I assume for source scenario we also were ending up with a partially substituted extension type

That's correct. The test showed such partially substituted extension type, but we also had an error for the declaration which prevented successful compilation.